### ISSUE TITLE
Question about the Ladv loss

### ISSUE THREAD

**Issue by lauraset** [wangle53/CSA-CDGAN#3]
Hi, @wangle53. The Ladv loss in `train.py`  seems wrong.
For the generator, it aims to generate the fake image similar to the real one.
Therefore, the Ladv should be `err_d_fake = l_bce(pred_fake, real_label)`

**Comment by wangle53** [comment]
Thanks for your attention. In BCE loss function, its common foumulation is −(ylog(p(x)+(1−y)log(1−p(x)). Here if the label of a sample is '0', we use (1−y)log(1−p(x)) to calculate its loss. In our paper, to be constant with the BCE loss fuction used in our codes, we formulate Ladv loss as -log(1-D(G(I))).

**Comment by lauraset** [comment]
Thanks for your nice help. I got it. Your work inspires me a lot.

Best regards,
Yinxia Cao



---- Replied Message ----
| From | ***@***.***> |
| Date | 09/07/2022 23:03 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [wangle53/CSA-CDGAN] Question about the Ladv loss (Issue #3) |

Thanks for your attention. In BCE loss function, its common foumulation is −(ylog(p(x)+(1−y)log(1−p(x)). Here if the label of a sample is '0', we use (1−y)log(1−p(x)) to calculate its loss. In our paper, to be constant with the BCE loss fuction used in our codes, we formulate Ladv loss as -log(1-D(G(I))).

**Comment by lauraset** [comment]
Hi, @wangle53. It still confuses me.  For the `err_d_fake = l_bce(pred_fake, fake_label)` for optimizing generator, it should be `err_d_fake = l_bce(pred_fake, real_label)`, since the generator aims to generate the fake image similar to the real one.  The following code is from `train.py`. See the gan file at (https://github.com/eriklindernoren/PyTorch-GAN/blob/master/implementations/gan/gan.py). Thank you in advance.
```python
            #forward

            fake = netg(x)
            pred_real = netd(gt)
            pred_fake = netd(fake).detach()
            err_d_fake = l_bce(pred_fake, fake_label)
            err_g = l_con(fake, gt)
            err_g_total = ct.G_WEIGHT*err_g + ct.D_WEIGHT*err_d_fake
            
            pred_fake_ = netd(fake.detach())
            err_d_real = l_bce(pred_real, real_label)
            err_d_fake_ = l_bce(pred_fake_, fake_label)
            err_d_total = (err_d_real + err_d_fake_) * 0.5
```

**Comment by wangle53** [comment]
I understand what you mean and I think you are absolutely right. However, we experimentd the setting of err_d_fake = l_bce(pred_fake, real_label) and found the performence is not good enough. Therefore, the loss if set to err_d_fake = l_bce(pred_fake, fake_label). In this paper, the proposed model is not the same as GAN. Only adversial learning is employed to enhance the model performance.

**Comment by lauraset** [comment]
Hi, @wangle53. Thank you for this detailed reply. I got it.

**Comment by wangle53** [comment]
You're welcome. Best wishes.

**Comment by 1998wq** [comment]
hello
